### PR TITLE
creating parent map function + flexible augmentations 

### PR DESCRIPTION
**Pull request by AvivSham**
Hi,
How are you? 
I have added your code a few tricks.
1) augmentation can be performed in 2 ways - (1) composition of all the available augmentation functions (2) randomly selecting one augmentation function and apply it.
2) more modular code - instead of calling several times for ```data.map``` I create you a "parent" map function, basically, this is the convention.

you just need to modify the description of ```DataLaoder``` class.
I hope that this merge suits you :)

### CODE REVIEW

**Comment by HasnainRaz** [comment]
Hey Aviv,

Thank you for the contribution, looks great! Let me just give it a try since I haven't touched Tensorflow in a year so I'm unsure of how the random.sample python function will interact with the TF functions. I will merge it once I'm done testing.

Thanks again for the contrib!

**Comment by AvivSham** [comment]
I'm happy to contribute :)
can you please drop here your email? maybe we can collaborate on research.

File: dataloader.py [hunk 1]
```diff
@@ -142,7 +149,36 @@ def _one_hot_encode(self, image, mask):
...
+                image, mask = self._corrupt_saturation(image, mask)
+                image, mask = self._crop_random(image, mask)
+                image, mask = self._flip_left_right(image, mask)
+            else:
+                options = [self._corrupt_brightness,
+                           self._corrupt_contrast,
+                           self._corrupt_saturation,
+                           self._crop_random,
+                           self._flip_left_right]
+                augment_func = random.choice(options)
```
Comment (HasnainRaz) [inline dataloader.py]: It seems TF records the choice the first time the sampling happens (if random crop was sampled, then it is always applied since the graph of the pipeline is constructed on the first run). We probably need to add tf.cond variables to apply random transforms for each image

File: dataloader.py [hunk 2]
```diff
@@ -142,7 +149,36 @@ def _one_hot_encode(self, image, mask):
...
+                image, mask = augment_func(images_path, masks_path)
```
Comment (HasnainRaz) [inline dataloader.py]: should be augment_func(image, mask) as we can't augment the path

**Review by HasnainRaz — CHANGES_REQUESTED** [review]

**Comment by HasnainRaz** [comment]
Sure Aviv, get in touch if you have anything interesting: hasnain3257 at gmail

File: dataloader.py (hunk 2 above)
Comment (AvivSham) [reply dataloader.py]: please check the last commit I have changed it.

File: dataloader.py (hunk 1 above)
Comment (AvivSham) [reply dataloader.py]: It looks like TF compiles the map function to a static graph if that is the case then the first selected augmentation function will be always used. To solved this scenario I have decorated the map function with ```tf.function``` which knows how to deal with py functions like random and compile it. 
I tried several times and it works, please do your tests and merge :)

File: dataloader.py [hunk 3]
```diff
@@ -109,8 +116,8 @@ def _resize_data(self, image, mask):
         """
         Resizes images to specified size.
         """
-        image = tf.image.resize(image, [self.image_size, self.image_size])
-        mask = tf.image.resize(mask, [self.image_size, self.image_size], method='nearest')
+        image = tf.image.resize(image, self.image_size)
+        mask = tf.image.resize(mask, self.image_size)
```
Comment (HasnainRaz) [inline dataloader.py]: TF requires image size to be a tuple (class accepts integer), so this needs to be a tuple, or the docstring needs to change

Comment (HasnainRaz) [inline dataloader.py]: For resizing masks, we should always use the nearest neighbour method, as masks should have integer values, and bilinear interpolation will result in float values.

File: dataloader.py [hunk 4]
```diff
@@ -142,7 +149,38 @@ def _one_hot_encode(self, image, mask):
         
         return image, one_hot_map
 
-    def data_batch(self, batch_size, augment, shuffle=False, one_hot_encode=False):
+    @tf.function
```
Comment (HasnainRaz) [inline dataloader.py:155]: I like this solution!

**Review by HasnainRaz — CHANGES_REQUESTED** [review]
Minor corrections needed

File: dataloader.py (hunk 3 above)
Comment (AvivSham) [reply dataloader.py]: when you call the dataloader class you will pass the size as tuple for example:
dataloader = Dataloader("/usr/local/images", "/usr/local/masks", (256,256))
that is fine :)

Comment (AvivSham) [reply dataloader.py]: BTW, default input argument shouldn't be mutable hence I changed the channels argument from `[3,3]` to `(3,3)`.  
Please check the new commit you can test it out but don't forget to pass image size as a tuple I also documented it so it should "scream" at you if you won't do it.

**Review by HasnainRaz — APPROVED** [review]
LGTM, there is a minor bug, but I will correct it
